src/Validator.js

```diff
@@ -72,6 +62,89 @@ class Validator {
     this.clientRepSocket = null;
     this.currentMultisig = new Multisig();
     this.electionTimeout = conf.electionTimeout;
+    this.randomizedLeadershipPoolInterval = this.electionTimeout / 3 + (Math.random() * this.electionTimeout) / 3;
+
+    this.selfNodeInfo = {
+      id: uuidV4(),
+      pid: process.pid,
+      startedAt: new Date(),
+      networkInfo: {
+        leaderPubAddr: this.pubAddr,
+        leaderRepAddr: this.repAddr,
+        leaderClientRepAddr: this.clientRepAddr,
+      },
+    };
+
+    logger.debug(`Creating new validator instance. Node info: `, JSON.stringify(this.selfNodeInfo));
+
+    // etcd setup
+    this.EtcdConstructor = typeof this.EtcdConstructor === 'function' ? this.EtcdConstructor : Etcd3;
```

Comment:
> why do we need this ternary here?

That's what makes the constructor parameter optional. If you want to inject your own Etcd client you can, but otherwise it falls back to just using the default library which is what 99.9% of the time will people want (my own estimate).

Reply:
The `options.EtcdConstructor` parameter. You can use it to inject the Etcd client of your choice (like a mock for tests).

Reply:
ah, I see that `Object.assign(this, options);`